plugins/openai/src/llm.ts

```diff
@@ -0,0 +1,669 @@
+    const delta = choice.delta;
+
+    if (delta.tool_calls) {
+      // check if we have functions to calls
+      for (const tool of delta.tool_calls) {
+        if (!tool.function) {
+          continue; // oai may add other tools in the future
+        }
+
+        if (tool.function.name) {
+          this.#toolCallId = tool.id;
+          this.#fncName = tool.function.name;
+          this.#fncRawArguments = tool.function.arguments || '';
+        } else if (tool.function.arguments) {
+          this.#fncRawArguments += tool.function.arguments;
+        }
+
+        if (this.#toolCallId && tool.id && tool.id !== this.#toolCallId) {
+          return this.#tryBuildFunction(id, choice);
+        }
+      }
+    }
+
+    if (
+      choice.finish_reason &&
+      ['tool_calls', 'stop'].includes(choice.finish_reason) &&
+      this.#toolCallId
+    ) {
+      // we're done with the tool calls, run the last one
+      return this.#tryBuildFunction(id, choice);
+    }
+
+    return {
+      requestId: id,
+      choices: [
+        {
+          delta: { content: delta.content || undefined, role: llm.ChatRole.ASSISTANT },
+          index: choice.index,
+        },
+      ],
+    };
+  }
+
+  #tryBuildFunction(
+    id: string,
+    choice: OpenAI.ChatCompletionChunk.Choice,
+  ): llm.ChatChunk | undefined {
+    if (!this.fncCtx) {
+      this.#logger.warn('oai stream tried to run function without function context');
+      return undefined;
+    }
+
+    if (!this.#toolCallId) {
+      this.#logger.warn('oai stream tried to run function but toolCallId is not set');
+      return undefined;
+    }
+
+    if (!this.#fncRawArguments || !this.#fncName) {
+      this.#logger.warn('oai stream tried to run function but rawArguments or fncName are not set');
+      return undefined;
+    }
+
+    const functionInfo = llm.oaiBuildFunctionInfo(
+      this.fncCtx,
+      this.#toolCallId,
+      this.#fncName,
+      this.#fncRawArguments,
+    );
+    this.#toolCallId = this.#fncName = this.#fncRawArguments = undefined;
+    this._functionCalls.push(functionInfo);
+
+    return {
+      requestId: id,
+      choices: [
+        {
+          delta: {
+            content: choice.delta.content || undefined,
+            role: llm.ChatRole.ASSISTANT,
+            toolCalls: this._functionCalls,
+          },
+          index: choice.index,
+        },
+      ],
+    };
+  }
+}
+
+const buildMessage = async (msg: llm.ChatMessage, cacheKey: any) => {
+  const oaiMsg: Partial<OpenAI.ChatCompletionMessageParam> = {};
+
+  switch (msg.role) {
+    case llm.ChatRole.SYSTEM:
+      oaiMsg.role = 'system';
+      break;
+    case llm.ChatRole.USER:
+      oaiMsg.role = 'user';
+      break;
+    case llm.ChatRole.ASSISTANT:
+      oaiMsg.role = 'assistant';
+      break;
+    case llm.ChatRole.TOOL:
+      oaiMsg.role = 'tool';
+      if (oaiMsg.role === 'tool') {
+        oaiMsg.tool_call_id = msg.toolCallId;
+      }
```

Comment:
```suggestion
      oaiMsg.tool_call_id = msg.toolCallId;
```

Reply:
this is for typescript validations, it doesn't work without this type enforcement

Reply:
Ah lol ok, weird that it isn't smart enough since the previous line literally set it to tool